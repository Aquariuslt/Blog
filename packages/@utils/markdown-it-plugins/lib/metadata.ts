import * as yaml from 'js-yaml';
import {format} from 'date-fns';

const METADATA_RE = /```metadata\n([\s\S]*?)```/g;

function parse(meta: string): BlogModel.Metadata {
  return yaml.load(meta);
}

function detectMetadata(md): void {
  md.block.ruler.before('fence', 'metadata', (state, startLine, endLine): boolean => {
    const startPos = state.bMarks[startLine] + state.tShift[startLine];
    const maxPos = state.eMarks[endLine];
    const measureBlock = state.src.slice(startPos, maxPos);

    const matches = METADATA_RE.exec(measureBlock);

    if (!matches) {
      return false;
    }


    state.src = state.src.replace(METADATA_RE, '');

    if (state.env) {
      const matchContent = matches[1];
      const metadata = parse(matchContent);

      metadata.created = format(metadata.created, 'YYYY-MM-DD');
      metadata.updated = format(metadata.updated, 'YYYY-MM-DD');


      state.env.metadata = metadata;
    }

    return false;
  });
}


export default detectMetadata;
