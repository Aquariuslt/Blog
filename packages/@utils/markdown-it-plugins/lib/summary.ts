const DEFAULT_SUMMARY_LENGTH = 100;

function detectSummary(md, options?) {
  md.core.ruler.push('detect_summary', function (state) {
    let summary = '';
    let summaryLength = DEFAULT_SUMMARY_LENGTH;

    if (options && options.len) {
      summaryLength = options.len;
    }

    let tokens = state.tokens;


    tokens.forEach((token, index) => {
      if (index > 0 && token.type === 'inline'
        && tokens[index - 1].type === 'paragraph_open') {
        if (summary.length < summaryLength) {
          summary += token.content.substr(0, summaryLength - summary.length);
        }
      }
    });

    if (state.env) {
      state.env.summary = summary;
    }
  });
}

export default detectSummary;
