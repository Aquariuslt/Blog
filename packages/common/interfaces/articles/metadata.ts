export interface ArticleMetadata {
  id: string;
  title: string;
  description: string;
  created: string;
  updated: string;
  categories: string[];
  tags: string[];
  cover: string;
}
