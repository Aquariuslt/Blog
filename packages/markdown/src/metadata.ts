import * as matter from 'gray-matter';

export const metadata = (md: string) => matter(md).data;
export const source = (md: string) => matter(md).content;
