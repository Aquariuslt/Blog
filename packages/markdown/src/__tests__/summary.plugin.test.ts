import * as MarkdownIt from 'markdown-it';
import { SummaryPlugin } from '../summary.plugin';
import { source } from '../metadata';
import { read } from '../__tests__/test.fixtures.helper';

describe('markdown-it plugin: summary', () => {
  it('# should get summary with more than 120 chars without any options', () => {
    const md = MarkdownIt().use(SummaryPlugin);
    const raw = source(read(`sample-article.md`));
    const context = {};

    md.parse(raw, context);
    expect(context).toHaveProperty('summary');
    expect(context['summary'].length).toBeGreaterThanOrEqual(120);
    expect(context['summary']).toMatchSnapshot();
  });

  it('# should get summary with spec chars with options.len', () => {
    const options = {
      len: 50
    };
    const md = MarkdownIt().use(SummaryPlugin, options);
    const raw = source(read(`sample-article.md`));
    const context = {};

    md.parse(raw, context);
    expect(context).toHaveProperty('summary');
    expect(context['summary'].length).toBeGreaterThanOrEqual(options.len);
  });

  it('# should get no summary when no context', () => {
    const md = MarkdownIt().use(SummaryPlugin);
    const raw = source(read(`sample-article.md`));
    md.parse(raw, undefined);
  });
});
