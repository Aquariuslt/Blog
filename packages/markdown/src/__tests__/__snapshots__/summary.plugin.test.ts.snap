// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown-it plugin: summary # should get summary with more than 120 chars without any options 1`] = `
"This is an example markdown content with all one-pass test cases.This section should load as summary in parsed data.Getting started snippet
"
`;
